renderdoc/driver/shaders/dxbc/dxbc_debug.h

```diff
@@ -147,6 +147,31 @@ struct GlobalState
   std::map<SampleEvalCacheKey, ShaderVariable> sampleEvalCache;
 };
 
+class DebugAPIWrapper
+{
+public:
+  virtual void SetCurrentInstruction(uint32_t instruction) = 0;
+  virtual void AddDebugMessage(MessageCategory c, MessageSeverity sv, MessageSource src,
+                               std::string d) = 0;
+
+  virtual bool CalculateRcp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateRsq(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateExp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateLog(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateSinCos(const ShaderVariable &input, ShaderVariable &output1,
+                               ShaderVariable &output2) = 0;
+
+  virtual ShaderVariable GetSampleInfo(const DXBC::ASMOperation &op) = 0;
+  virtual ShaderVariable GetBufferInfo(const DXBC::ASMOperation &op) = 0;
+  virtual ShaderVariable GetResourceInfo(const DXBC::ASMOperation &op, uint32_t mipLevel,
+                                         int &dim) = 0;
```

Comment:
I'd rather not 'leak' the whole `ASMOperation` to this external function, and have it be responsible for knowing which operand is needed from which opcode. From what I can see in these `Get*Info` functions the op is only used (aside from the error string) to determine the binding - UAV/SRV and binding number, or perhaps rasterizer in one case. I'd rather do that operation-specific decode in the debug code itself then call out through this interface just specifying the binding alone.

Reply:
To clarify, are you suggesting combining these Get*Info functions into a single call, similar to the math functions? I took a preliminary look, and it feels like the majority of work would be branched, so leaving it as separate functions makes more sense to me. I agree about distilling the params down to strictly what is needed, making that change now.

I left some of the work for these opcodes in dxbc_debug.cpp (such as handling sample patterns, swizzling, etc.) since I couldn't think of a reason why D3D11/D3D12 would need to diverge there. Would you prefer making the API wrapper handle all of it, but moving this common functionality to a shared location, outside of dxbc_debug.cpp?

Reply:
I don't mean we should combine these into a single call - I agree with you that the implementations would mostly be separate anyway. What I mean is e.g. looking at `GetSampleInfo` it has this code:

```
  else if(op.operands[1].type == DXBC::TYPE_RESOURCE && op.operands[1].indices.size() == 1 &&
          op.operands[1].indices[0].absolute && !op.operands[1].indices[0].relative)
```

which is a bit of a leaky abstraction - now each implementation of `DebugAPIWrapper` needs to know not only how to map the binding and get the sample info data, but also needs to know that specifically operand 1 is the one to look at, not operand 0. Similarly in `GetResourceInfo` it needs to look up operand 2.

What I'm suggesting is that for these functions the small amount of code that parses/decodes the op should be kept in dxbc_debug.cpp and it should just pass through e.g. an enum for rasterizer/srv/uav and a binding number.

Reply:
Understood and fixed.